Add Concentrate Last order sorting

BlueSoapTurtle opened the pull request:
Since the concentrate modifier requires you to spam click, this option sorts Concentrate potions last to reduce accidental miss-clicks when processing multiple orders.

This fixes #79 

This is a very small change adding the following PotionOrderSorting enum:
```java
CONCENTRATE_LAST("Concentrate Last", Comparator.comparing((PotionOrder order) ->
                order.potionModifier() == PotionModifier.CONCENTRATED ? 1 : 0)
        .thenComparing(order -> order.potionModifier().ordinal()));
```

Wondering if we can change it to go in the order 
Crystalize>Homogenise>Concentrate as it's currently doing 
Homogenise>Crystalize>Concentrate 
Which forces you to walk further than you should? Please give input on this

raiyni commented:
I think this could be added to BY_STATION maybe, unless there is some crazy pathing strat? And you can just compare by `alchemyObject`

Huntardedd commented:
![image](https://github.com/user-attachments/assets/79a3efdb-99ee-4164-af0e-172f22991b1e)

Appreciate the quick response and work towards an update.
Not sure if implemented yet, or added to "by station" but currently still ordering concentrate in the middle of potion orders list. 
Even if not implemented, seems relevant as the first post mentions the order should currently be "Homogenise>Crystalize>Concentrate".

BlueSoapTurtle commented:
@Huntardedd Thanks for the suggestion, and appreciate the feedback! 
I've changed the existing BY_STATION sorting option to prioritize the order as CRYSTALISED > HOMOGENOUS > CONCENTRATED. I also made it sort on the potion type alphabetically so similar orders will be in order, might help a bit! 

@raiyni I removed the new order and instead changed BY_STATION to order them, 
Could you clarify your comment about using alchemyObject for ordering?

raiyni commented:
> @Huntardedd Thanks for the suggestion, and appreciate the feedback! I've changed the existing BY_STATION sorting option to prioritize the order as CRYSTALISED > HOMOGENOUS > CONCENTRATED. I also made it sort on the potion type alphabetically so similar orders will be in order, might help a bit!
> 
> @raiyni I removed the new order and instead changed BY_STATION to order them, Could you clarify your comment about using alchemyObject for ordering?
> 
> Also, the change is not implemented yet, this PR has to be approved and merged first, here's a photo of it in action though ![image](https://private-user-images.githubusercontent.com/6739300/381230696-665fd2ee-06eb-4af4-96c6-88613f4fc2d0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAyNTAwMjMsIm5iZiI6MTczMDI0OTcyMywicGF0aCI6Ii82NzM5MzAwLzM4MTIzMDY5Ni02NjVmZDJlZS0wNmViLTRhZjQtOTZjNi04ODYxM2Y0ZmMyZDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMzBUMDA1NTIzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTQ4YzFmODk4MzJlZGI5MGQwNjliODY3ZGY4NzFlOWYyNDdjNGZkYzE2ZmM5YmMxY2RlNDY1ZWRhZGRlMzMyYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.5Xzekj579nn8YeKBCVnJ_OO7Bs9baAKzPe_C9hPPtUc)

Using that field to sort them instead of the ordinal, the stations are placed in order of the object Id's

Notloc commented:
There is a special case in the movement efficiency if you're inclined to handle it.

If the order only requires the homogenizing and concentrating stations, the optimal route is concentrate -> homogenize.
(7 ticks vs 5 ticks)

All other cases seem to be optimal with the currently proposed sorting.

Notloc commented:
Wait no, crystalize -> concentrate -> homogenize is 7 tick.
The proposed order is 8 tick.

Notloc commented:
I should read more carefully.
I see now that the main motivation is to reduce misclicks when concentrating potions.
I'll do a separate PR to handle an optimal route sorting option instead.

BlueSoapTurtle commented:
@raiyni Using the alchemyObject instead of the modifier results in
CRYSTALISED/ALEMBIC > HOMOGENOUS/AGITATOR > CONCENTRATED/RETORT 
instead of 
HOMOGENOUS/AGITATOR > CONCENTRATED/RETORT > CRYSTALISED/ALEMBIC
But it's still not clear that it uses that order, so I think it's better to clearly show that in the comparator

@Notloc Just saw the other PR, nice!

I'm feeling a bit included to restore the old patch and have a separate sorting to put the CONCENTRATED/RETORT last in the order, and then having it be CRYSTALISED/ALEMBIC > HOMOGENOUS/AGITATOR > CONCENTRATED/RETORT 
which would be achieved by using the alchemyObject and then we don't touch the By station one?

Alternatively we add a config for setting station priority and it uses those to figure out what to use, not changing them would result in the vanilla layout and then you can change the order as needed, lmk what people think

hex-agon commented on src/main/java/work/fking/masteringmixology/PotionOrderSorting.java:
```diff
@@ -4,7 +4,22 @@
     VANILLA("Vanilla (random)", null),
-    BY_STATION("By station", Comparator.comparing(order -> order.potionModifier().ordinal()));
+    // Sort by modifier, in the order CRYSTALISED > HOMOGENOUS > CONCENTRATED
+    BY_STATION("By station", Comparator
```
Create a `PotionComparators` class then create a static method that returns the comparator and use method references in the enum like:
`BY_STATION("By station", PotionComparators::byStation)`

BlueSoapTurtle replied:
Had to do PotionComparators.byStation() otherwise it expects an implementation of the Comparator, rebased on master as well!

Huntardedd commented:
> I'm feeling a bit included to restore the old patch and have a separate sorting to put the CONCENTRATED/RETORT last in the order, and then having it be CRYSTALISED/ALEMBIC > HOMOGENOUS/AGITATOR > CONCENTRATED/RETORT which would be achieved by using the alchemyObject and then we don't touch the By station one?
> 
> Alternatively we add a config for setting station priority and it uses those to figure out what to use, not changing them would result in the vanilla layout and then you can change the order as needed, lmk what people think

As of today, re-visiting the minigame and plugin, both vanilla and by station still operating the same as a week ago showing Concentrate(retort) in the middle of potion orders list.
uninstalled and reinstalled plugin, no new order sorting option or setting station config as mentioned above.
Both "issue" and "pull request" threads showing as closed, and seeing multiple "approved changes"/"checks passed".
Curious as to when/if these changes are implemented?

Huntardedd commented:
> I'm feeling a bit included to restore the old patch and have a separate sorting to put the CONCENTRATED/RETORT last in the order, and then having it be CRYSTALISED/ALEMBIC > HOMOGENOUS/AGITATOR > CONCENTRATED/RETORT which would be achieved by using the alchemyObject and then we don't touch the By station one?
> 
> Alternatively we add a config for setting station priority and it uses those to figure out what to use, not changing them would result in the vanilla layout and then you can change the order as needed, lmk what people think

As of today, revisiting the minigame and plugin, both "vanilla" and "by station" order sorting options operating same as a week ago.
Uninstalled and reinstalled plugin, refreshed orders list, no new order option in order sorting selection, or station priority config option as mentioned above.
Currently ordering concentrate(retort) in the middle of potion orders list on "vanilla", as well as ordering Concentrate(retort) first on "by station"
![image](https://github.com/user-attachments/assets/aaebaf0c-6563-413b-aa94-384b8954ca56)
![image](https://github.com/user-attachments/assets/33c9ff6a-c7c1-4eeb-9aae-64bf4df6b260)

Both "issue" and "pull request" threads showing as closed, and seeing multiple "approved changes"/"checks passed".
Curious as to when/If these changes are, or will be implemented?
